bundles/org.openhab.ui/web/src/components/widgets/system/oh-colorpicker.vue

```diff
@@ -0,0 +1,115 @@
+<template>
+  <div ref="container" style="width: 100%"></div>
+</template>
+
+<script>
+import mixin from '../widget-mixin'
+
+export default {
+  name: 'oh-colorpicker',
+  mixins: [mixin],
+  data () {
+    return {
+      colorpicker: null,
+      delayCommand: false,
+      delayUpdate: false,
+      pendingCommand: null,
+      pendingUpdate: null,
+      init: false
+    }
+  },
+  mounted () {
+    if (this.color) {
+      this.initColorpicker()
+    }
+  },
+  beforeDestroy () {
+    this.colorpicker.destroy()
+  },
+  computed: {
+    color () {
+      const state = this.context.store[this.config.item].state
+      if (state.split(',').length === 3) {
+        let color = this.context.store[this.config.item].state.split(',')
+        color[0] = parseInt(color[0])
+        color[1] = color[1] / 100
+        color[2] = color[2] / 100
+        return color
+      }
+      return null
+    }
+  },
+  watch: {
+    color (val) {
+      if (this.colorpicker) {
+        this.updateValue(val)
+      } else {
+        this.initColorpicker()
+      }
+    }
+  },
+  methods: {
+    initColorpicker () {
+      const vm = this
+      this.colorpicker = this.$f7.colorPicker.create(Object.assign({}, this.config, {
+        containerEl: this.$refs.container,
+        modules: this.config.modules || ['hsb-sliders'],
+        value: {
+          hsb: this.color
+        },
+        on: {
+          change (colorpicker, value) {
+            // skip the first update
+            if (!vm.init || vm.context.store[vm.config.item].state === '-') {
+              vm.init = true
+              return
+            }
+            if (!value.hsb) return
+            vm.sendCommand(value.hsb)
+          }
+        }
+      }))
+    },
+    sendCommand (hsb) {
+      this.pendingUpdate = [...hsb]
+      this.pendingCommand = [...hsb]
+      const state = this.commandFromHSB(hsb)
+      if (state !== this.context.store[this.config.item].state) {
+        if (!this.delayCommand) {
+          this.delayCommand = true
+          this.delayUpdate = true
+          this.$store.dispatch('sendCommand', { itemName: this.config.item, cmd: state })
+          setTimeout(() => {
+            const pendingCommand = [...this.pendingCommand]
+            this.pendingCommand = null
+            this.delayCommand = false
+            if (pendingCommand != null) {
+              this.sendCommand(pendingCommand)
+            }
+          }, 200)
+        }
+        setTimeout(() => {
+          this.updateValue(this.pendingUpdate)
+          this.delayUpdate = false
+        }, 2000)
+      }
+    },
+    commandFromHSB (hsb) {
```

Comment:
Question about common functions like this, I could see many widgets needing to do similar conversions, should there be a mix-in or common class that would contain common conversion like this?  For example, i can see both a switch and slider needing to share common util functions as both can operate on a Dimmer item and may need to convert ON, OFF, 0 , 100, etc..... I'm more curious what your opinion here is for others (like me).

Reply:
Yes you're absolutely right, there's some refactoring to do in the widgets and in other places to share common code, mixins might be a good choice... there's a widget-mixin already which does the evaluation of expressions in the config and a few other things, there could be some utility functions in there, or maybe another mixin dedicated to conversions.